Addressed HeatExchanger:AirToAir:SensibleAndLatent negative effectiveness and sizing

Nigusse opened the pull request:
Pull request overview
---------------------
 - Fixes #7901
 - Fixed Heat Exchanger nominal volume flow rate sizing issue. Now the nominal air flow rate is sized to the maximum or the minimum outdoor design air flow rate depending on the OA controller specification. Now the defect file runs successfully.
 - Provided warning when there is incorrect user-inputs for latent effectiveness at 100% and 75% rated flow rates. Warning message is thrown when only one of them is zero.  

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport

### Reviewer
This will not be exhaustively relevant to every PR.
 - [x] Perform a Code Review on GitHub
 - [x] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [x] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] ~If feature, test running new feature, try creative ways to break it~
 - [x] CI status: all green or justified
 - [x] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally

Commit:
HX Sized to Maximum Vol Flow Rate

Commit:
added warning for incorrect latent eff input

Nigusse commented:
Sample warning error message:

```
   ** Warning ** HeatExchanger:AirToAir:SensibleAndLatent "AIR-TO-AIR HX 15"
   **   ~~~   ** ...Latent Effectiveness at 100% Cooling Air Flow = 0.00
   **   ~~~   ** ...Latent Effectiveness at 75% Cooling Air Flow = 0.73
   **   ~~~   ** ...Latent cooling effectiveness at 75% and 100% rated flow can be 0.0 but not one of them only.
```

Nigusse commented:
Sample results of HX nominal supply air flow rate sizing:

Before the fix:
```
 Component Sizing Information, Controller:OutdoorAir, CONTROLLER OUTDOOR AIR 15, Maximum Outdoor Air Flow Rate [m3/s], 4.88937
 Component Sizing Information, HeatExchanger:AirToAir:SensibleAndLatent, AIR-TO-AIR HX 15, Design Size Nominal Supply Air Flow Rate [m3/s], 0.48682
```

After the fix:

```
 Component Sizing Information, Controller:OutdoorAir, CONTROLLER OUTDOOR AIR 15, Maximum Outdoor Air Flow Rate [m3/s], 4.88937
 Component Sizing Information, HeatExchanger:AirToAir:SensibleAndLatent, AIR-TO-AIR HX 15, Design Size Nominal Supply Air Flow Rate [m3/s], 4.88937
```

Nigusse commented:
Three test suite files (ASHRAE9012016_HotelLarge_Denver.idf, DOAS_wNeutralSupplyAir_wFanCoilUnits.idf, and DOAToUnitVentilator.idf) have failed. This fix is expected to show diffs because of sizing results change. See below eio output for ASHRAE9012016_HotelLarge_Denver.idf test file:

```
  Component Sizing Information, Controller:OutdoorAir, VAV WITH REHEAT_OA_CONTROLLER, Maximum Outdoor Air Flow Rate [m3/s], 26.10562
  Component Sizing Information, Controller:OutdoorAir, VAV WITH REHEAT_OA_CONTROLLER, Minimum Outdoor Air Flow Rate [m3/s], 11.78914
- Component Sizing Information, HeatExchanger:AirToAir:SensibleAndLatent, VAV WITH REHEAT HEAT RECOVERY, Design Size Nominal Supply Air Flow Rate [m3/s], 11.78914
+ Component Sizing Information, HeatExchanger:AirToAir:SensibleAndLatent, VAV WITH REHEAT HEAT RECOVERY, Design Size Nominal Supply Air Flow Rate [m3/s], 26.10562
```

Nigusse commented:
Five test suite files are expected to show diffs because of changing the nominal supply air flow rate change from minimum to maximum design outdoor air flow rate for "HeatExchanger:AirToAir:SensibleAndLatent".

mjwitte commented:
Hmmm, so that does raise the question of which is the correct airflow rate to size the HX. In some applications, the min OA flow might be correct, because anything > min OA could mean economizer operation so the HX would be off/bypassed. But for a DOAS (which I think was the defect file application), with variable flow, Max OA flow makes more sense. Can we detect this and make the right choice, or does the user need an input to control?

Nigusse commented:
The test suite file DOAToPTAC.idf has HR HeatExchanger on the OA system. This is DOAS, so the maximum and minimum are the same. See sizing results below: 

```
Component Sizing Information, Controller:OutdoorAir, DOAS OA CONTROLLER, Maximum Outdoor Air Flow Rate [m3/s], 0.26908
 Component Sizing Information, Controller:OutdoorAir, DOAS OA CONTROLLER, Minimum Outdoor Air Flow Rate [m3/s], 0.26908
 Component Sizing Information, HeatExchanger:AirToAir:SensibleAndLatent, DOAS HEAT RECOVERY, Design Size Nominal Supply Air Flow Rate [m3/s], 0.26908

```

Nigusse commented:
I am trying to think what kind of design condition limit the HR-HX flow to the minimum other than DOAS, in which case the minimum and the maximum design outdoor flows are the same?

Nigusse commented:
One scenario I can think of is when the Economizer Control Type in controlled OA controller is set to *NoEconomizer*. If this is the case, then I will investigate to mine that information to set a flag that can be used during sizing such the heat exchanger is sized to minimum flow. Is there any other scenario where the HX needs to be sized to minimum flow?

mjwitte commented:
But a variable flow DOAS could have different min/max OA flows. I don't think we have any example files like that, but it can be done. It may work to check the economizer and HX lockout settings.

Nigusse commented:
OK. Below are the four cases the HX nominal flow is sized to Minimum OA Floe If there is outdoor air controller in the OA system:

- Economizer Type is *NoEconmizer*
- Economizer Control Action Type is set to *MinimumFlowWithBypass*
- When the Lockout Type is *LockoutWithHeating* or *LockoutWithCompressor* 
- When the Lockout Type is *NoLockout* and Heat Recovery Bypass Control Type is *BypassWhenOAFlowGreaterThanMinimum*

Nigusse commented:
@mjwitte I think, when the Economizer Type not NoEconomizer but the Heat Recovery Bypass Control Type is BypassWhenWithinEconomizerLimits, then the HX should be sized to Minimum Flow?

Commit:
added option for HR sizing to minimum OA flow

Commit:
Added unit tests for HR nominal flow autosize

Nigusse commented on src/EnergyPlus/DataSizing.cc:
```diff
@@ -313,6 +313,7 @@ namespace DataSizing {
     int DataDesicDehumNum(0);                       // index to desiccant dehumidifier
     bool DataDesicRegCoil(false);                   // TRUE if heating coil desiccant regeneration coil
     bool HRFlowSizingFlag(false);                   // True, if it is a heat recovery heat exchanger flow sizing
+    bool SizeHRHXtoMinFlow(false);                  // true, if the heat recovery heat exchanger nominal flow is sized to min OA flow
```
New flag variable is used to decide the of use minimum or maximum OA flow for sizing Heat Recovery HX.

Nigusse commented on src/EnergyPlus/HeatRecovery.cc:
```diff
@@ -529,6 +532,12 @@ namespace HeatRecovery {
+                ShowContinueError("..." + cNumericFieldNames(3)  + " = " + RoundSigDigits(rNumericArgs(3), 2));
+                ShowContinueError("..." + cNumericFieldNames(5)  + " = " + RoundSigDigits(rNumericArgs(5), 2));
+                ShowContinueError("...Latent heating effectiveness at 75% and 100% rated flow can be 0.0 but not one of them only.");
+            }
```
Throws warning when only one of the heating latent effectiveness (at 100% or 75% of rated flow) user-input value is zero. But allows both to be zero.

mjwitte replied:
It's possible that a user might want one of these to be zero without the other, so I'm not sure this it the most general warning. The real problem is if the calculated effectiveness (either sensible or latent) goes negative. Would it be better to check for effectiveness <0 during the simulation, clip it to zero and throw a one-time warning there that lists the pertinent inputs (like you have here) and then says 
"... effectiveness is less than zero. Check the following inputs. Effectiveness reset to zero and the simulation continues." ?

Protecting against negative effectiveness is crucial, but I don't see a code change to add that.

Nigusse commented on src/EnergyPlus/HeatRecovery.cc:
```diff
@@ -543,6 +552,13 @@ namespace HeatRecovery {
+                ShowContinueError("..." + cNumericFieldNames(9) + " = " + RoundSigDigits(rNumericArgs(9), 2));
+                ShowContinueError("...Latent cooling effectiveness at 75% and 100% rated flow can be 0.0 but not one of them only.");
+            }
+
```
Same as the above but this is for cooling.

Nigusse commented on src/EnergyPlus/HeatRecovery.cc:
```diff
@@ -5386,6 +5403,28 @@ namespace HeatRecovery {
+        }
+        return HXNominalFlowSizingFlag;
+    }
 } // namespace HeatRecovery
```
Added new function that set the value of SizeHRHXtoMinFlow flag depending on the OA controller inputs. Various combination of economizer inputs such economizer control type, bypass control, and lockout status determine the flag value.

mjwitte replied:
If don't see a check on the HX field "Economizer Lockout". If this is set to No, then HXNominalFlowSizingFlag = false because the HX will always be operating and economizer settings don't matter.

mjwitte replied:
If `thisOAController.Econo == MixedAir::NoEconomizer`, I don't think it will ever send a bypass signal, so I think this should be the first test and if it's true, set the function to false and skip the rest.
Why do the Lockout options apply here?

Nigusse commented on src/EnergyPlus/ReportSizingManager.cc:
```diff
@@ -2630,26 +2630,26 @@ namespace ReportSizingManager {
+                            }
+                            if (SizeHRHXtoMinFlow) {
+                                if (FinalSysSizing(CurSysNum).DesOutAirVolFlow > 0.0) {
+                                    AutosizeDes = FinalSysSizing(CurSysNum).DesOutAirVolFlow;
```
The HR heat exchanger nominal flow rate is autosized to maximum OA flow if the SizeHRHXtoMinFlow is false and the heat exchanger is on OA system.

Nigusse replied:
This change does not affect HR heat exchanger nominal flow sizing if the HR unit is on air loop main branch.

mjwitte replied:
This isn't really sizing to max OA flow when SizeHRHXtoMinFlow is false. It's assuming that the max OA flow rate will match the CurDuctType flow or more likely `DesMainFlow`. 
Is there a FinalSysSizing variable for max OA flow?
Is `CurDutcType` even applicable and current when `CurOASysNum` is >0? I know this if else block is present in the current code, but I'm suspicious that it ever lands anywhere but the final else. @rraustad ?

Nigusse commented:
Current fix shows when the economizer is NoLockout and the Heat Recovery Bypass Control Type is *BypassWhenOAFlowGreaterThanMinimum*, the HR is sized at max flow but when the Heat Recovery Bypass Control Type is *BypassWhenWithinEconomizerLimits*, the HR is sized to min flow. I think this logic should be reversed. @mjwitte what do you think?

Commit:
revised a logic for HR flow sizing

Nigusse commented:
Below is the logic that sets a flag used for sizing heat recovery heat exchanger nominal flow to minimum OA value:

```
        if (thisOAController.Econo == MixedAir::NoEconomizer || thisOAController.EconBypass ||
            thisOAController.Lockout == MixedAir::LockoutWithHeatingPossible ||
            thisOAController.Lockout == MixedAir::LockoutWithCompressorPossible ||
            (thisOAController.Lockout == MixedAir::NoLockoutPossible &&
             thisOAController.HeatRecoveryBypassControlType == DataHVACGlobals::BypassWhenOAFlowGreaterThanMinimum)) {
             HXNominalFlowSizingFlag = true;}
```

Nigusse commented:
This pull request is ready for review.

mjwitte commented on src/EnergyPlus/DataSizing.hh:
```diff
@@ -314,6 +314,7 @@ namespace DataSizing {
     extern int DataDesicDehumNum;                     // index to desiccant dehumidifier
     extern bool DataDesicRegCoil;                     // TRUE if heating coil desiccant regeneration coil
     extern bool HRFlowSizingFlag;                     // True, if it is a heat recovery heat exchanger flow sizing
+    extern bool SizeHRHXtoMinFlow;                    // true, if the heat recovery heat exchanger nominal flow is sized to min OA flow
```
I'm not a fan of adding yet another DataSizing flag, but I'll hold that thought for now.

Nigusse replied:
I know it complicates but I do not if there is any better was to do when you have all kinds control flavors.

mjwitte commented on src/EnergyPlus/ReportSizingManager.cc:
```diff
@@ -2630,26 +2630,26 @@ namespace ReportSizingManager {
                 } else if (SizingType == SystemAirflowSizing) {
                     if (HRFlowSizingFlag) { // HX sizing
                         if (CurOASysNum) {
-                            if (FinalSysSizing(CurSysNum).DesOutAirVolFlow > 0.0) {
```
Nit-picking, but this could simply be a one-line change here.
`if(SizeHRHXtoMinFlow && FinalSysSizing(CurSysNum).DesOutAirVolFlow > 0.0) {`

Nigusse replied:
OK. will do the change. Done.

mjwitte commented on src/EnergyPlus/HeatRecovery.cc:
```diff
@@ -5386,6 +5403,28 @@ namespace HeatRecovery {
 
+    bool GetHeatRecoveryHXNominalFlowSizingFlag(int const OASysNum) 
+    {
+        bool HXNominalFlowSizingFlag;
```
The name of this function and variable is somewhat ambiguous - does "nominal" mean MinOA? At a minimum this needs a comment to explain what this function is doing. I'd probably rename the function to "GetHeatRecoveryHXMinFlowSizingFlag"?

Nigusse replied:
Agreed, will do the change.

Nigusse replied:
done.

Nigusse replied:
Renamed the local variable to HXMinFlowSizingFlag and re-arranged the logic.

mjwitte reviewed (commented):
Some initial code review comments/questions.

Commit:
renamed the function GetHeatRecoveryHXMinFlowSizingFlag

Nigusse commented:
Only OA controller object has max OA flow rate but it gets from the same set of assignments that we use in sizing. See below:

```
                    if (SELECT_CASE_var == ControllerOutsideAir) {

                        CheckSysSizing(CurrentModuleObject, this->Name);

                        {
                            auto const SELECT_CASE_var1(CurDuctType);
                            if (SELECT_CASE_var1 == Main) {
                                this->MaxOA = FinalSysSizing(CurSysNum).DesMainVolFlow;
                            } else if (SELECT_CASE_var1 == Cooling) {
                                this->MaxOA = FinalSysSizing(CurSysNum).DesCoolVolFlow;
                            } else if (SELECT_CASE_var1 == Heating) {
                                this->MaxOA = FinalSysSizing(CurSysNum).DesHeatVolFlow;
                            } else if (SELECT_CASE_var1 == Other) {
                                this->MaxOA = FinalSysSizing(CurSysNum).DesMainVolFlow;
                            } else {
                                this->MaxOA = FinalSysSizing(CurSysNum).DesMainVolFlow;
                            }
                        }

                    } else if (SELECT_CASE_var == ControllerStandAloneERV) {
```

Nigusse commented:
Below is code snippet that sets the flag used to size the HX to minimum flow. The first two and the fourth cases, I think, are OK. However, the third option is debatable but I made this logic based on the documentation. Especially the *LockoutWithCompressorPossible* option the economizer is locked out during heating and cooling operation, so I thought the heat exchange should be sized to the minimum but in the case of *LockoutWithHeatingPossible* the economizer is locked during heating only. I am OK if you want to remove the later. Any ways, I am happy to be corrected if any of the logic are not applicable.    
```

        if (thisOASys.OAControllerIndex > 0) {
            auto &thisOAController(MixedAir::OAController(thisOASys.OAControllerIndex));
            if (thisOAController.ControllerType_Num == MixedAir::ControllerOutsideAir) {
                if (thisOAController.Econo == MixedAir::NoEconomizer) {
                    HXMinFlowSizingFlag = true;
                } else if (thisOAController.EconBypass) {
                    HXMinFlowSizingFlag = true;
                } else if (thisOAController.Lockout == MixedAir::LockoutWithHeatingPossible ||
                           thisOAController.Lockout == MixedAir::LockoutWithCompressorPossible) {
                    HXMinFlowSizingFlag = true;
                } else if (thisOAController.Lockout == MixedAir::NoLockoutPossible &&
                           thisOAController.HeatRecoveryBypassControlType == DataHVACGlobals::BypassWhenOAFlowGreaterThanMinimum) {
                    HXMinFlowSizingFlag = true;
                }
            }
        }
```

Commit:
various cleanup

Commit:
Merged develop; resolved conflict.

mjwitte commented:
```
 if (thisOAController.Econo == MixedAir::NoEconomizer) {
                    HXMinFlowSizingFlag = false;
```
Because the HX will never be bypassed if there is no economizer. I guess for all of these we need to look at the OA controller logic and see under what conditions is can set the HX bypass flag to true.

Nigusse commented:
@mjwitte regarding providing warning when the effectiveness values goes negative, the code [here](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/HeatRecovery.cc#L2195-L2196) never lets the effectiveness to drop below zero. It looks negative effectiveness value is not used in the calculation. Do you still want to see the warning message added?

Nigusse commented:
My understanding has always been when NoEconomizer is the Economizer Type, then the OA system provides just ventilation air requirement, i.e. the minimum OA amount. Note that this logic applies only when the HX is on Outdoor Air System. It is for this reason that I set the HXMinFlowSizingFlag  = true for this case. 

If the HX is on the main branch, then the HX is sized whatever flow that applicable to the main branch. All these new logic added apply only when the HX is on Outdoor Air System.

mjwitte commented:
RE: negative effectiveness, when I posted the issue, I saw a negative effectiveness in the debugger which was causing the bad results. Not sure how it got there. If you can't see where that was happening I'll try to reproduce it.

RE: NoEconomizer. That's true for a recirculating system with some fraction of outdoor air. But it's not true for a DOAS with variable flow. Maybe we need some other way to figure out the intent. OR maybe we just ditch all of this guessing the intent and add a new field somewhere in Sizing:System or in the HX object to declare the sizing intent? I'd rather not do that, but that may be the most straightforward approach.

Nigusse commented:
Re: Negative Effectiveness: I removed the previous warning message based on user input and replaced it with warning message provided only when the actual effectiveness is less than zero. Now a onetime warning will be thrown when the actual effectiveness is negative for sensible and latent calculations in heating and cooling mode. I agree it is worth to throw warning if it some how results in negative effectiveness.

Commit:
Modified HX sizing logic and warning message

Commit:
Merge branch 'develop' into 172492479_Issue7901

Nigusse commented:
@mjwitte Addressed most of your review change request except some of the economizer setting impact on the logic of the HX sizing.

Commit:
Merge branch 'develop' into 172492479_Issue7901

Commit:
Merge branch 'develop' into 172492479_Issue7901

Commit:
Rolled back HX sizing change and modified unit tests

Nigusse commented:
@mjwitte Rolled back the heat recovery HX flow sizing changes and modified the unit tests accordingly.

mjwitte commented on src/EnergyPlus/HeatRecovery.cc:
```diff
@@ -2190,11 +2192,52 @@ namespace HeatRecovery {
+                    onetimewarningsensibleEff = true;
+                }
+            }
+            bool static onetimewarninglatentEff = false;
```
Rather than a static bool here, I'd add a couple of bools to the `ExchCond` struct. That way if more than one HX exhibits the problem, they can all be reported.

Nigusse replied:
OK. I will address your change request. @mjwitte Thanks.

Nigusse replied:
Also adding the following HX Air Volume Flow ratio as part of the error message. The flow ratio used to calculate the effectiveness.

```cpp 
ShowContinueError("...Heat Exchanger Air Volume Flow Ratio = " + RoundSigDigits(HXAirVolFlowRatio, 2));```

mjwitte reviewed (commented):
@Nigusse Tested this with the original defect file. The new error message shows up (good), but the file still fails with a fatal error. Running this in the debugger reveals that effectiveness is calculated again in lines 2308-2323 and it's coming up negative again here. So it needs to be trapped here as well.

Commit:
Merge branch 'develop' into 172492479_Issue7901

Commit:
Addressed review comments

mjwitte reviewed (approved):
Original defect file now runs to completion, showing the new warning. Nice thorough unit tests. All good.
